### ISSUE TITLE
Initializing an `action.Configuration` is not threadsafe

### ISSUE THREAD

**Issue by qsymmachus** [helm/helm#9089]
**tl;dr** Initializing an `action.Configuration` is not threadsafe due to underlying calls to `apiextv1.AddToScheme` and `apiextv1beta1.AddToScheme`, which modify maps without memory synchronization.

---

My team maintains a tool that uses Helm to deploy services in our prod k8s cluster. The tool often performs deploys that consist of multiple Helm releases. When done synchronously, this can take some time, so we tested performing each release in a separate goroutine.

In the wild, this resulted in `concurrent map writes` errors. I believe this is due to [how Helm initializes action configs](https://github.com/helm/helm/blob/v3.4.1/pkg/action/action.go#L365):

```go
func (c *Configuration) Init(getter genericclioptions.RESTClientGetter, namespace, helmDriver string, log DebugLog) error {
	kc := kube.New(getter)
	kc.Log = log
        // etc
}
```

The call to `kube.New` adds [CRDs to the scheme](https://github.com/helm/helm/blob/v3.4.1/pkg/kube/client.go#L76), which is not threadsafe, because it modifies (ordinary, non-threadsafe) maps within the `Scheme` struct. This is called out in [documentation for the type](https://github.com/kubernetes/apimachinery/blob/v0.19.3/pkg/runtime/scheme.go#L45):

> Schemes are not expected to change at runtime and are only threadsafe after registration is complete.

We can't avoid multiple concurrent calls to `Init` because our releases may target different namespaces. We've worked around this by reimplementing `Init` in our own codebase, where our version of `kube.New` skips adding CRDs to the scheme. We do this only once when the service starts.

As you can imagine this isn't a sustainable long-term solution because our reimplemented functions will have to be updated with each upgrade of Helm. But we'd like to keep performing concurrent Helm releases because it's provided a significant performance boost for our end users.

My question for the Helm team is, **why are CRDs added to the scheme with every single call to `kube.New`?** My understanding is these additions only need to happen once. Is there a way we could rewrite action config initialization to avoid these repeated additions, or make them optional? 

Thanks in advance for your time!

**Comment by bacongobbler** [comment]
As you pointed out, adding the CRDs to the scheme is wrapped in a `sync.Once` call.

From https://golang.org/pkg/sync/#Once.Do:

> if once.Do(f) is called multiple times, only the first call will invoke f, even if f has a different value in each invocation. A new instance of Once is required for each function to execute.

Can you verify this with a test case?

**Comment by qsymmachus** [comment]
@bacongobbler you pointed me in exactly the right direction. My examples all refer to the latest version, but I should have clarified that we are currently using `v3.0.2`. In that version, `kube.New` does not wrap the Scheme additions in a `sync.Once`: https://github.com/helm/helm/blob/v3.0.2/pkg/kube/client.go#L65

So this problem has already been solved, we just need to upgrade to `v3.4.1` to reap the benefits. Thanks again for your help!
